### PR TITLE
Fix/refresh test status

### PR DESCRIPTION
**Pull request by jjelschen**
**Note:** 
[web-workspace-navigator/integration/src/app/test.execution.service.mock.ts](https://github.com/test-editor/web-workspace-navigator/blob/792a19e80d740cc4a8525456712a5c1cc5e92034/integration/src/app/test.execution.service.mock.ts) is just a copy of [web-workspace-navigator/src/demo/app/test.execution.service.mock.ts](https://github.com/test-editor/web-workspace-navigator/blob/master/src/demo/app/test.execution.service.mock.ts), which is less than ideal for all the usual reasons. There is heavy code duplication between the actual library code, the demo app, and the integration app, in general, which I think we need to find a better solution for at some point…

### CODE REVIEW

File: src/lib/src/component/navigation/navigation.component.spec.ts [hunk 1]
```diff
@@ -758,4 +759,75 @@ it('emits "navigation.open" message when the enter key is pressed', () => {
...
+      new Map<string, ElementState>([[tclFile.path, ElementState.LastRunFailed]])));
+    when(executionService.status(tclFile.path)).thenReturn(Promise.resolve(
+      new Response(new ResponseOptions({ body: 'FAILED' }))));
+
+    // when
+    component.refresh();
+
+    // then
+    fixture.whenStable().then(() => {
+      let updatedTclFile = component.getWorkspace().getElement('subfolder/file.tcl');
```
Comment (gunther-bachmann) [inline src/lib/src/component/navigation/navigation.component.spec.ts]: at this point I have no idea that tclFile.path == 'subfolder/file.tcl'. using tclFile.path might be confusing too. Maybe put a 'let updatedTclFilePath = tclFile.path;' in the given section and use that?

File: src/lib/src/component/navigation/navigation.component.spec.ts [hunk 2]
```diff
@@ -758,4 +759,75 @@ it('emits "navigation.open" message when the enter key is pressed', () => {
...
+      expect(updatedTclFile).not.toBe(tclFile);
+      expect(component.getWorkspace().getElement('subfolder/file.tcl').state).toEqual(ElementState.LastRunFailed);
```
Comment (gunther-bachmann) [inline src/lib/src/component/navigation/navigation.component.spec.ts]: why not use updatedTclFile here?

File: src/lib/src/component/navigation/navigation.component.spec.ts [hunk 3]
```diff
@@ -758,4 +759,75 @@ it('emits "navigation.open" message when the enter key is pressed', () => {
...
+  it('stops all polling for test status on component destruction', fakeAsync(() => {
+    // given
+    setupWorkspace(component, fixture);
+    component.selectElement(tclFile.path);
+    fixture.detectChanges();
+    let responseDelayMillis = 10;
+    setAsyncTestStatusServiceRunningResponse(executionService, responseDelayMillis);
+    component.run();
+    tick(responseDelayMillis);
+    verify(executionService.status(tclFile.path)).twice();
```
Comment (gunther-bachmann) [inline src/lib/src/component/navigation/navigation.component.spec.ts]: twice? can't see in this test why this should happen twice. maybe just add a comment?

File: src/lib/src/component/navigation/navigation.component.spec.ts [hunk 4]
```diff
@@ -758,4 +759,75 @@ it('emits "navigation.open" message when the enter key is pressed', () => {
...
+    // when
+    component.refresh();
+
+    // then
+    tick(responseDelayMillis);
+    verify(executionService.status(tclFile.path)).twice();
+    let updatedTclFile = component.getWorkspace().getElement(tclFile.path);
+    expect(updatedTclFile).not.toBe(tclFile);
+    expect(updatedTclFile.state).toEqual(ElementState.Running);
+    component.ngOnDestroy();
```
Comment (gunther-bachmann) [inline src/lib/src/component/navigation/navigation.component.spec.ts]: destroy is probably necessary because of 'component.run()' in the given part. Maybe add a comment, which part of the 'then' is actually something like a teardown

File: src/lib/src/component/navigation/navigation.component.ts [hunk 5]
```diff
@@ -12,20 +12,23 @@ import { ElementState } from '../../common/element-state';
...
   styleUrls: ['./navigation.component.css']
 })
 
-export class NavigationComponent implements OnInit {
+export class NavigationComponent implements OnInit, OnDestroy {
   static readonly HTTP_STATUS_CREATED = 201;
   static readonly NOTIFICATION_TIMEOUT_MILLIS = 4000;
 
   private workspace: Workspace;
+  private stopPolling: Subject<void> = new Subject<void>();
```
Comment (gunther-bachmann) [inline src/lib/src/component/navigation/navigation.component.ts]: wouldn't stopPollingTestStatus be a better name?

File: src/lib/src/service/execution/test.execution.service.ts [hunk 6]
```diff
@@ -24,9 +26,23 @@ export class TestExecutionService {
     return this.http.get(this.getURL(path, TestExecutionService.statusURLPath) + '&wait=true').toPromise();
   }
 
+  statusAll(): Promise<Map<string, ElementState>> {
```
Comment (gunther-bachmann) [inline src/lib/src/service/execution/test.execution.service.ts:29]: 'statusAll' is a bit misleading here, since not all test status are returned, only the ones the server cares to provide (afaik). I can't think of a better name right now, but maybe you have an idea?

File: src/lib/src/component/navigation/navigation.component.ts [hunk 7]
```diff
@@ -46,17 +49,36 @@ export class NavigationComponent implements OnInit {
...
+      this.updateTestStates();
       return this.workspace;
     }).catch(() => {
       this.errorMessage = 'Could not retrieve workspace!';
       return undefined;
     });
   }
 
+  private updateTestStates(): void {
+    this.stopPolling.next();
```
Comment (gunther-bachmann) [inline src/lib/src/component/navigation/navigation.component.ts]: I don't immediately get why the polling has to be stopped here. Is the subscription refreshed (only in setWorkspace)? Subscription, polling and stop polling seem to need some comments, so that (even) I understand their working ...

File: src/lib/src/component/navigation/navigation.component.test.setup.ts [hunk 8]
```diff
@@ -60,21 +69,18 @@ export function setTestExecutionServiceResponse(service: TestExecutionService, s
...
-export function setTestStatusServiceResponse(service: TestExecutionService, statusCode: number) {
-  const responseBeforeTermination = new Response(new ResponseOptions({
-    status: statusCode,
-    body: 'RUNNING'
-  }));
-  const responseAfterTermination = new Response(new ResponseOptions({
-    status: statusCode,
-    body: 'SUCCESS'
-  }));
+export function setTestStatusServiceResponse(service: TestExecutionService) {
```
Comment (gunther-bachmann) [inline src/lib/src/component/navigation/navigation.component.test.setup.ts]: 'set' does not seem to be the best prefix here, what do you think? Maybe something like 'mockTestStatusServiceWithRunningRunningSuccessSequence'? It is quite long, but might help understand its working.

File: src/lib/src/component/navigation/navigation.component.test.setup.ts [hunk 9]
```diff
@@ -60,21 +69,18 @@ export function setTestExecutionServiceResponse(service: TestExecutionService, s
...
   when(service.status(tclFile.path))
       .thenReturn(Promise.resolve(responseBeforeTermination))
       .thenReturn(Promise.resolve(responseBeforeTermination))
       .thenReturn(Promise.resolve(responseAfterTermination));
 }
 
+export function setAsyncTestStatusServiceRunningResponse(service: TestExecutionService, delayMillis: number) {
```
Comment (gunther-bachmann) [inline src/lib/src/component/navigation/navigation.component.test.setup.ts]: 'set' ? (see above)
maybe 'mockTestStatusServiceWithPromiseRunning'

File: src/lib/src/component/navigation/navigation.component.spec.ts (hunk 1 above)
Comment (jjelschen) [reply src/lib/src/component/navigation/navigation.component.spec.ts]: Absolutely right, I introduced a new variable to make it clearer, but I chose to name it `pathInWorkspaceToBeRefreshed` (because the path remains the same, while the object identities of the corresponding `WorkspaceElement` instances are expected to change; `updatedTclFilePath` kind of carries with it the notion that it is only the path of the element _after_ the refresh).

File: src/lib/src/component/navigation/navigation.component.spec.ts (hunk 2 above)
Comment (jjelschen) [reply src/lib/src/component/navigation/navigation.component.spec.ts]: No good reason…

File: src/lib/src/component/navigation/navigation.component.ts (hunk 7 above)
Comment (jjelschen) [reply src/lib/src/component/navigation/navigation.component.ts]: Right, that's a bit convoluted, I guess. The gist of it is, when the workspace is refreshed, it gets thrown out completely. Unfortunately, the `Workspace` class is not well encapsulated, at all, so there can, and therefore _will_ be `WorkspaceElement` instances that remain referenced somewhere after a refresh. In particular, the polling tasks update the test state using a direct reference to the corresponding workspace element. Therefore, they should be stopped on a refresh, because they will a) cause traffic senselessly as the status updates won't be visible, and b) so the references to those workspace elements are removed and garbage collection can reclaim them (avoiding a memory leak).

… I actually started to refactor this before I implemented it the way it is now, trying to better encapsulate class `Workspace`, and refreshing its state rather than throwing it out, completely. That quickly proved to be much more involved than I had hoped, though, so I stopped following that road. I think it would still make sense to refactor this, as it will only get worse…

File: src/lib/src/service/execution/test.execution.service.ts (hunk 6 above)
Comment (jjelschen) [reply src/lib/src/service/execution/test.execution.service.ts:29]: Well, the server is supposed to return the state of all tests it knows about, so that's the notion of 'all' here. Its basically the status of all tests except those that have never been executed, which, one could argue, don't really have a status. That's pretty much how it is implemented in the backend, while the frontend uses `IDLE` as the state for anything it doesn't know better, i.e. it is that state of tests that have never been run, but also the state of non-executable files and folders…

Long story short, I can't really think of a concise, yet more descriptive name here, and I don't think its all that misleading.
Maybe we [c|sh]ould add some documentation to our interfaces / REST endpoints – here, and in general – so the intended contract / semantics are made more explicit(?)
Problem is, we probably don't want to maintain documentation in (at least) two spots, backend and frontend.
